pxr/usd/usdGeom/triangulation.h

```diff
@@ -0,0 +1,140 @@
+    /// and \r represents the next index to \p.
+    FacePoint(const GfVec3f& p, const GfVec3f& q, const GfVec3f& r);
+
+    FacePoint(FacePoint const& other);
+
+    /// Computes the distance to the half plane <q, p, r> and checks whether the distance is
+    /// positive or negative.
+    bool HasSameOrientation(const FacePoint& other) const;
+
+    /// Computes two half planes <s, q, p> and <s, q, r> and checks whether those two planes
+    /// faces each other.
+    bool IsVisible(const GfVec3f& s) const;
+
+    /// Whether the point is one of the points consisting this FacePoint.
+    bool HasPoint(const GfVec3f& point) const;
+
+    /// Computes three half planes <q, p, r>; <p, q, r>; <r, p, q> and checks whether a point lies
+    /// in all of the three planes. Equivalent to find if a point lies within a triangle.
+    bool ContainsPoint(const GfVec3f& s) const;
+private:
+    GfVec3f m_p;
+    GfVec3f m_q;
+    GfVec3f m_r;
+    GfVec3d m_direction;
+};
+
+/// \class Face
+/// A face is a representation of a Face in a Mesh. It consist of single public method: FanTriangulate.
+class Face
+{
+public:
+    Face(const VtVec3fArray& points,
+        const VtIntArray& indices,
+        const size_t indexStart,
+        const size_t vertexCount);
+
+    /// Returns true if it is possible to triangulate a face, false otherwise.
+    /// If it is possible it will return an array of indices and faces to replace.
+    bool FanTriangulate(VtIntArray& indices, VtIntArray& faces) const;
+private:
+    size_t size() const
+    {
+        return m_vertexCount;
+    }
+
+    GfVec3f const& operator[](const size_t index) const
+    {
+        size_t offset = m_indices[m_indexStart + index];
+        return m_points[offset];
+    }
+
+    bool IsValid() const
+    {
+        return m_vertexCount >= 3;
+    }
+
+    FacePoint GetReference() const;
+
+    /// Returns true if the current face is a convex shape, false otherwise.
+    /// Checking if a face is convex is a linear operation and if it is convex,
+    /// the original indices and faces will not suffer modifications.
+    bool IsConvex() const;
+
+    /// Returns true if the current face is a star shaped, false otherwise.
+    /// A polygon is star shaped if there is at least one point where all points are visible.
+    /// Checking if a face is star shaped is a linear operation and if it is star shaped,
+    /// the original indices will suffer a shift.
+    bool IsStarShaped(const FacePoint& reference, size_t& index) const;
+
+    /// Returns true if the current face is ear clipping, false otherwise.
+    /// A simple polygon should be able to triangulate. This will modify all indices and face counts.
+    /// This implementation is O(n * r) where r are the number of concave vertices.
+    bool IsEarClipping(const FacePoint& reference, VtIntArray& indices, VtIntArray& faceCount) const;
+
+    void add(VtIntArray& indices, const size_t index) const
```

Comment:
I suppose this is a utility function? Maybe document it as such, or make it protected if it's internal only?

Also Add would be consistent casing for the name